lib/ukvmem/vmem.c

```diff
@@ -0,0 +1,1106 @@
+}
+
+static struct uk_vma *vmem_vma_try_merge(struct uk_vma *vma)
+{
+	vmem_vma_try_merge_with_next(vma);
+	return vmem_vma_try_merge_with_prev(vma);
+}
+
+static int vmem_vma_split(struct uk_vma *vma, __vaddr_t vaddr,
+			  struct uk_vma **new_vma)
+{
+	struct uk_vma *v = __NULL;
+	int rc;
+
+	UK_ASSERT(vma);
+	UK_ASSERT(new_vma);
+
+	UK_ASSERT(vaddr >= vma->start && vaddr < vma->end);
+	UK_ASSERT(PAGE_Lx_ALIGNED(vaddr, MAX(vma->page_lvl, PAGE_LEVEL)));
+	UK_ASSERT(PAGE_Lx_ALIGNED(vma->end - vaddr,
+		  MAX(vma->page_lvl, PAGE_LEVEL)));
+
+	rc = VMA_SPLIT(vma, vaddr, &v);
+	if (unlikely(rc))
+		return rc;
+
+	if (!v) {
+		v = uk_malloc(vma->vas->a, sizeof(struct uk_vma));
+		if (unlikely(!v))
+			return -ENOMEM;
+	}
+
+	UK_ASSERT(v);
+
+	UK_INIT_LIST_HEAD(&v->vma_list);
+	v->start	= vaddr;
+	v->end		= vma->end;
+	v->vas		= vma->vas;
+	v->ops		= vma->ops;
+	v->attr		= vma->attr;
+	v->flags	= vma->flags;
+	v->page_lvl	= vma->page_lvl;
+	v->name		= vma->name;
+
+	vma->end	= vaddr;
+
+	uk_list_add(&v->vma_list, &vma->vma_list);
+
+	*new_vma = v;
+	return 0;
+}
+
+static inline int vmem_vma_need_split(struct uk_vma *vma, __sz len,
+				      unsigned long *attr)
+{
+	UK_ASSERT(vma);
+
+	return (len > 0) && (!attr || (vma->attr != *attr));
+}
+
+static int vmem_vma_split_vmas(struct uk_vas *vas, __vaddr_t vaddr, __sz len,
+			       unsigned long *attr, struct uk_vma **start,
+			       struct uk_vma **end, int strict)
+{
+	struct uk_vma *vma_start = __NULL, *vma_end, *vma;
+	__vaddr_t vend;
+	__sz left, right;
+	int rc;
+
+	UK_ASSERT(vas);
+	UK_ASSERT(start);
+	UK_ASSERT(end);
+
+	/*
+	 * We can have the following cases:
+	 *
+	 * AAAAAAAAA        AAAAAAAAA         AAAAAAAAA         AAAAAAAAA
+	 * ^-------^        ^---^                 ^---^           ^---^
+	 *
+	 * AAA C BBB        AAA C BBB         AAA C BBB         AAA C BBB
+	 * ^-------^         ^------^         ^------^           ^-----^
+	 */
+	rc = vmem_vma_find_range(vas, &vaddr, &len, &vma_start, &vma_end,
+				 strict);
+	if (unlikely(rc))
+		return rc;
+
+	vend  = vaddr + len;
+	left  = vaddr - vma_start->start;
+	right = vma_end->end - vend;
+
+	if (vmem_vma_need_split(vma_start, left, attr)) {
+		rc = vmem_vma_split(vma_start, vaddr, &vma);
+		if (unlikely(rc))
+			return rc;
+
+		UK_ASSERT(vma);
+
+		if (vma_start == vma_end)
+			vma_end = vma;
+
+		vma_start = vma;
+	}
+
+	if (likely(len > 0)) {
+		if (vmem_vma_need_split(vma_end, right, attr)) {
+			rc = vmem_vma_split(vma_end, vend, &vma);
+			if (unlikely(rc)) {
+				vmem_vma_try_merge(vma_start);
+				return rc;
+			}
+
+			UK_ASSERT(vma);
+		}
+	} else {
+		UK_ASSERT(vma_start == vma_end);
+	}
+
+	*start = vma_start;
+	*end   = vma_end;
+
+	return 0;
+}
+
+int vma_op_deny()
+{
+	return -EPERM;
+}
+
+int vma_op_unmap(struct uk_vma *vma, __vaddr_t vaddr, __sz len)
+{
+	UK_ASSERT(vaddr >= vma->start);
+	UK_ASSERT(vaddr + len <= vma->end);
+
+	return ukplat_page_unmap(vma->vas->pt, vaddr, len / PAGE_SIZE, 0);
```

Comment:
Just a question: `len % PAGE_SIZE == 0` is always true? Otherwise, the round down behavior might be wrong.

Reply:
This function should only be used as default operation in which case it is always called by `vmem_vma_unmap()`. This function has extensive length checks. But as a good measure, I added another `UK_ASSERT(PAGE_ALIGNED(len));`.